bundle/edu.gemini.model.p1/src/main/scala/edu/gemini/model/p1/immutable/Proposal.scala

```diff
@@ -115,6 +115,16 @@ case class Proposal(meta:Meta,
 
   def resetObservationMeta:Proposal = copy(observations = observations.map(_.copy(meta = None)))
 
+  // Calculate the program time as a ratio of total time.
+  def programTimeRatio: Double = {
+    def timeSum(extract: Observation => Option[TimeAmount]): TimeAmount =
+      TimeAmount.sum(observations.map(extract).flatten)
+    val progTime  = timeSum(_.progTime)
+    val partTime  = timeSum(_.partTime)
+    val totalTime = progTime |+| partTime
+    totalTime.isEmpty ? 1.0 | (progTime.hours / totalTime.hours)
```

Comment:
It could be empty in theory if no time was assigned to the observations. That being said, this method should never be called if that happens, but I wanted to eliminate any risk of dividing by zero.